File: 2016/Day_06/Python/solution.py
"""Solution for 2016 Day 5 -- Signals and Noise."""

from collections import Counter


def input_per_line(file: str):
    """This is for when each line is an input to the puzzle. The newline character is stripped."""
    with open(file, 'r') as input_file:
        return [line.rstrip() for line in input_file.readlines()]


def find_most_common_letters(word_list: list, word_length: int) -> list:
    """Take in a list of words, count the number of times each letter appears per position.

    :param word_list: the words to search.
    :param word_length: how long each of the words are
    :return: list of characters that appear most often in each position
    """
    counter_list = [Counter() for number in range(word_length)]
    for word in word_list:
        characters = [character for character in word]
        for position, character in enumerate(characters):
            counter_list[position][character] += 1
    return [counter_list[number].most_common(1)[0][0] for number in range(len(counter_list))]


if __name__ == '__main__':
    scrambled_signal = input_per_line("../input.txt")
    scrambled_word_length = len(scrambled_signal[0])
    most_common = find_most_common_letters(scrambled_signal, scrambled_word_length)
    print(f"The message is {''.join(most_common)}")

File: 2016/Day_06/Go/solution.go
// Solution for Advent of Code Day 06 - Signals and Noise

package main

import (
	"bufio"
	"fmt"
	"os"
)

// MultipleLines returns the input file if it contains multiple lines of text
func MultipleLines(fileName string) ([]string, error) {
	inputSlice := make([]string, 0)
	file, err := os.Open(fileName)
	if err != nil {
		return inputSlice, err
	}
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		inputSlice = append(inputSlice, scanner.Text())
	}
	err = file.Close()
	if err != nil {
		return inputSlice, err
	}
	return inputSlice, nil
}

// FindMaxLetter finds the letter that appears most often in the map
func FindMaxLetter(positionMap map[rune]int) rune {
	maxVal := 0
	var returnRune rune
	for key, value := range positionMap {
		if value > maxVal {
			maxVal = value
			returnRune = key
		}
	}
	return returnRune
}

func main() {
	var positionZero = make(map[rune]int)
	var positionOne = make(map[rune]int)
	var positionTwo = make(map[rune]int)
	var positionThree = make(map[rune]int)
	var positionFour = make(map[rune]int)
	var positionFive = make(map[rune]int)
	var positionSix = make(map[rune]int)
	var positionSeven = make(map[rune]int)
	counters := [8]map[rune]int{positionZero, positionOne, positionTwo, positionThree, positionFour, positionFive, positionSix, positionSeven}
	ourInput, _ := MultipleLines("/home/ermesa/Programming Projects/adventofcode/2016/Day_06/input.txt")
	for _, word := range ourInput {
		for position, character := range word {
			counters[position][character]++
		}
	}
	fmt.Println("The error-corrected message is:")
	for _, value := range counters {
		output := string(FindMaxLetter(value))
		fmt.Print(output)
	}
	fmt.Println("")
}
